pkg/kvstore/etcd.go

```diff
@@ -0,0 +1,372 @@
+//
+// Copyright 2016 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package kvstore
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/common/types"
+
+	client "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	ctx "golang.org/x/net/context"
+)
+
+type EtcdClient struct {
+	*client.Client
+	session     *concurrency.Session
+	sessionMU   *sync.RWMutex
+	lockPaths   map[string]*sync.Mutex
+	lockPathsMU *sync.Mutex
+}
+
+type EtcdLocker struct {
+	mutex     *concurrency.Mutex
+	localLock *sync.Mutex
+	path      string
+}
+
+func NewEtcdClient(config *client.Config, cfgPath string) (KVClient, error) {
+	var (
+		c   *client.Client
+		err error
+	)
+	if cfgPath != "" {
+		c, err = client.NewFromConfigFile(cfgPath)
+	} else if config != nil {
+		c, err = client.New(*config)
+	} else {
+		return nil, fmt.Errorf("empty configuration provided")
+	}
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Waiting for etcd client to be ready...")
+	s, err := concurrency.NewSession(c)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to contact etcd: %s", err)
+	}
+	log.Info("Etcd client ready")
+	ec := &EtcdClient{
+		Client:      c,
+		session:     s,
+		sessionMU:   &sync.RWMutex{},
+		lockPaths:   map[string]*sync.Mutex{},
+		lockPathsMU: &sync.Mutex{},
+	}
+	go func() {
+		for {
+			select {
+			case <-ec.session.Done():
+				newSession, err := concurrency.NewSession(c)
+				if err != nil {
+					log.Errorf("Error while renewing etcd session %s", err)
+					time.Sleep(10 * time.Second)
+				} else {
+					ec.sessionMU.Lock()
+					ec.session = newSession
+					ec.sessionMU.Unlock()
+					log.Debugf("Renewing etcd session")
+				}
+			}
+		}
+	}()
+	return ec, nil
+}
+
+func (e *EtcdClient) LockPath(path string) (KVLocker, error) {
+	e.lockPathsMU.Lock()
+	if e.lockPaths[path] == nil {
+		e.lockPaths[path] = &sync.Mutex{}
+	}
+	e.lockPathsMU.Unlock()
+
+	log.Debugf("Locking path %s", path)
+	// First we lock the local lock for this path
+	e.lockPaths[path].Lock()
+	e.sessionMU.RLock()
+	mu := concurrency.NewMutex(e.session, path)
```

Comment:
No, it doesn't make sense to use STM if you have that interface as it is. Perhaps it's ripe to be split apart and simplified in the future, detaching the pessimistic locking semantics from it.